framework/include/interfaces/FunctorInterface.h

```diff
@@ -0,0 +1,311 @@
+   * elemental data properly pre-initialized at the desired quadrature point
+   */
+  virtual T evaluate(const QpArg & qp, unsigned int state) const = 0;
+
+  /**
+   * @param tqp A pair with the first member corresponding to an \p ElementType, either Element,
+   * Neighbor, or Lower corresponding to the three different possible \p MooseVariableData
+   * instances. The second member corresponds to the desired quadrature point
+   * @return The requested element type data indexed at the requested quadrature point. As for the
+   * \p qp overload, there is a caveat: any variables involved in the functor evaluation must have
+   * their requested element data type properly pre-initialized at the desired quadrature point
+   */
+  virtual T evaluate(const std::tuple<Moose::ElementType, unsigned int, SubdomainID> & tqp,
+                     unsigned int state) const = 0;
+
+private:
+  /**
+   * clear cache data
+   */
+  void clearCacheData();
+
+  /// How often to clear the material property cache
+  std::set<ExecFlagType> _clearance_schedule;
+
+  // Data for traditional element-quadrature point property evaluations which are useful for caching
+  // implementation
+
+  /// Current quadrature point element (current cache key)
+  mutable dof_id_type _current_qp_elem = DofObject::invalid_id;
+
+  /// Current quadrature point element data (current cache value)
+  mutable std::vector<std::pair<bool, T>> * _current_qp_elem_data = nullptr;
+
+  /// Cached element quadrature point functor property evaluations. The map key is the element
+  /// id. The map values should have size corresponding to the number of quadrature points on the
+  /// element. The vector elements are pairs. The first member of the pair indicates whether a
+  /// cached value has been computed. The second member of the pair is the (cached) value. If the
+  /// boolean is false, then the value cannot be trusted
+  mutable std::unordered_map<dof_id_type, std::vector<std::pair<bool, T>>> _qp_to_value;
+};
+
+template <typename T>
+T
+FunctorInterface<T>::operator()(const Elem * const & elem, const unsigned int state) const
+{
+  return evaluate(elem, state);
+}
+
+template <typename T>
+T
+FunctorInterface<T>::operator()(const ElemAndFaceArg & elem_and_face,
+                                const unsigned int state) const
+{
+  return evaluate(elem_and_face, state);
+}
+
+template <typename T>
+T
+FunctorInterface<T>::operator()(const FaceArg & face, const unsigned int state) const
+{
+  return evaluate(face, state);
+}
+
+template <typename T>
+T
+FunctorInterface<T>::operator()(const QpArg & elem_and_qp, const unsigned int state) const
+{
+  if (_clearance_schedule.count(EXEC_ALWAYS))
+    return evaluate(elem_and_qp, state);
+
+  const auto elem_id = std::get<0>(elem_and_qp)->id();
+  if (elem_id != _current_qp_elem)
+  {
+    _current_qp_elem = elem_id;
+    _current_qp_elem_data = &_qp_to_value[elem_id];
```

Comment:
Have you done any profiling on this?  Do these map lookups show up?

Reply:
The cost of these lookups appear to be fairly trivial. This is profiling a 100k dof 3D calculation
```
         .          .    300:template <typename T>
         .          .    301:T
         .          .    302:FunctorInterface<T>::operator()(const ElemQpArg & elem_qp, const unsigned int state) const
         .          .    303:{
         .       70ms    304:  if (_clearance_schedule.count(EXEC_ALWAYS))
         .      2.65s    305:    return evaluate(elem_qp, state);
         .          .    306:
      10ms       10ms    307:  const auto elem_id = std::get<0>(elem_qp)->id();
      30ms       30ms    308:  if (elem_id != _current_qp_map_key)
         .          .    309:  {
         .          .    310:    _current_qp_map_key = elem_id;
         .       20ms    311:    _current_qp_map_value = &_qp_to_value[elem_id];
         .          .    312:  }
         .          .    313:  auto & qp_data = *_current_qp_map_value;
         .          .    314:  const auto qp = std::get<1>(elem_qp);
         .          .    315:
         .      3.70s    316:  return queryQpCache(qp, qp_data, elem_qp, state);
         .          .    317:}
```